### PR TITLE
Add stride-aware image profiling

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Add an opt-in `stride` argument to `profile()` so image-model callers can pass the desired target tensor and let THOP own efficient stride-sized profiling.

- Keeps every existing `profile(model, inputs=...)` caller unchanged.
- Uses one stride-sized profile for spatial-only models.
- Fits a second double-area profile when supported fixed-cost operations are present.
- Falls back to the supplied target input when proxy inputs are unsuitable.
- Bumps `ultralytics-thop` from 2.1.0 to 2.1.1.
- Aligns `AGENTS.md` Core Principles and exact-head PR workflow with `ultralytics/ultralytics`.

This lets Ultralytics remove its local FLOPs scaling and layer-inspection policy while preserving YOLOv3, YOLOv5, and third-party consumers of the existing API.

Deleted: duplicated image-size scaling and fallback ownership can now be removed from the Ultralytics consumer.

## Validation

- `python -m pytest tests/` — 20 passed.
- Ruff check and format — clean.
- Exact target comparison across all 66 Ultralytics model YAMLs — 66/66 matched, including classification, YOLO-World, and five RT-DETR models.
- Repeated real-model calls left no forward hooks behind.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🚀 Adds stride-aware MAC estimation for image models and releases THOP version 2.1.1.

### 📊 Key Changes

- Added an optional `stride` argument to `profile()` for estimating operations at a target image size using smaller, stride-aligned inputs.
- Supports integer or two-dimensional stride values with input validation.
- Uses one or two proxy profiles to account for both spatially scaling and size-independent operations, such as linear layers and adaptive pooling.
- Falls back to profiling the original target input when stride-based estimation is unsuitable or fails.
- Preserves existing profiling behavior when `stride` is omitted or layer-level details are requested.
- Added documentation and a regression test confirming stride-based results match direct target-size profiling.
- Updated the package version from `2.1.0` to `2.1.1`.
- Refined `AGENTS.md` development and review guidance around minimal changes, source-level fixes, and validation.

### 🎯 Purpose & Impact

- ⚡ Enables faster MAC estimation for large image inputs by profiling smaller representative tensors.
- 🎯 Improves accuracy for image models that combine spatial operations with fixed-cost components.
- 🔄 Maintains backward compatibility for existing `profile()` calls.
- 🛡️ Adds validation and fallback behavior to reduce the risk of incorrect estimates or unexpected failures.
- 📈 Helps users more efficiently assess model complexity during development and deployment.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/thop` 🚀 PR! This automated message helps with the initial review; an engineer will assist with the integration of your stride-aware profiling feature. Please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/thop/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/thop` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: thop/profile.py [hunk 1]
```diff
@@ -246,9 +247,50 @@ def dfs_count(module: nn.Module, prefix="\t") -> (int, int):
...
+        else:
+            if len(inputs) != 1 or not isinstance(inputs[0], torch.Tensor) or inputs[0].ndim != 4:
+                raise ValueError("stride requires inputs to contain one BCHW image tensor")
+            stride = (stride, stride) if isinstance(stride, int) else tuple(stride)
+            if len(stride) != 2 or min(stride) < 1:
+                raise ValueError("stride must be a positive integer or a pair of positive integers")
+
+            image = inputs[0]
+            target_height, target_width = image.shape[-2:]
+            fixed_ops = (nn.Linear, nn.AdaptiveAvgPool1d, nn.AdaptiveAvgPool2d, nn.AdaptiveAvgPool3d, nn.RNNBase)
```
Comment (UltralyticsAssistant) [inline thop/profile.py]: 💡 **MEDIUM**: The fixed-cost detection omits the supported `nn.RNNCell`, `nn.GRUCell`, and `nn.LSTMCell` hooks. These classes are registered above, but unlike `nn.RNN`/`nn.GRU`/`nn.LSTM` they do not derive from `nn.RNNBase`. A valid image wrapper such as `AdaptiveMaxPool2d(1) -> Flatten -> RNNCell` therefore takes the one-sample path and scales the cell's constant per-batch cost by image area, producing an incorrect target MAC count. Include the cell classes in the two-sample condition (and add coverage for one).

File: thop/profile.py [hunk 2]
```diff
@@ -246,9 +247,50 @@ def dfs_count(module: nn.Module, prefix="\t") -> (int, int):
...
+            required_samples = 2 if custom_ops or any(isinstance(m, fixed_ops) for m in model.modules()) else 1
+            samples = []
+            if target_height % stride[0] == target_width % stride[1] == 0 and stride[0] < target_height:
```
Comment (UltralyticsAssistant) [inline thop/profile.py]: 📝 **LOW**: This gate only permits proxy profiling when the target height exceeds the height stride, even though the probes vary width. For a valid wide input such as `(1, 3, 32, 640)` with `stride=32`, both stride-aligned proxy shapes are usable but this condition skips them and profiles the full target instead. This defeats the advertised fast path for rectangular inputs whose height is one stride; choose an axis that can be expanded or make the eligibility check account for the width-based probes.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The implementation preserves the existing path and cleans up temporary hooks correctly. However, the new fixed-cost classification misses supported recurrent cell modules, which can yield incorrect stride estimates, and its proxy eligibility unnecessarily bypasses the optimization for some rectangular inputs.

💬 Posted 2 inline comments

File: thop/profile.py (hunk 2 above)
Comment (glenn-jocher) [reply thop/profile.py]: Fixed in a233709 by gating on proxy area rather than target height. Verified exact parity for both 32x640 and 640x32 inputs locally.

File: thop/profile.py (hunk 1 above)
Comment (glenn-jocher) [reply thop/profile.py]: Fixed in a233709 by including RNNCell, GRUCell, and LSTMCell in THOP’s fixed-cost set. Kept committed coverage light as requested and verified an image-wrapper RNNCell estimate against the exact target profile locally.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you, @glenn-jocher, and congratulations on merging this impactful PR!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

Stride-aware MAC estimation, robust validation, fallback behavior, regression coverage, and the THOP 2.1.1 release make model profiling faster, more accurate, and fully backward compatible. This is a meaningful improvement for efficient development and deployment across image models—thank you for the thoughtful engineering! 🚀
